libzim/wrapper.pyx

```diff
@@ -406,15 +417,15 @@ cdef class FilePy:
         article = ReadArticle.from_read_article(art)
         return article
 
-    def get_metadata(self, name):
+    def get_metadata(self):
         """Get the file metadata.
         Returns
         -------
         dict
-            A dictionary with the file metadata
+            A dictionary with the standard metadata present on file
         """
-        article = self.get_article(f"M/{name}")
-        return article.content
+        return {name: bytes(self.get_article(f"M/{name}").content).decode("UTF-8")
+                for name in MANDATORY_METADATA_KEYS if self.has_article(f"M/{name}")}
```

Comment:
The new API will have a method called `get_metadata` that will return a specific metadata (given a name). It would be better to do the same here to avoid a new API change on python side.

Reply:
Ah, good to know! Should we return the decoded string directly here? Otherwise it's quite useless compared to `get_article()`…
Also, should we provide this dict under another name?

Reply:
It will be not possible to get metadata with `get_article` (or equivalent in new API).
I'm not sure that we need to decode the string all the time. From https://www.openzim.org/wiki/Metadata, all metadata in `M` namespace are strings so we could decode it. But the favicon, even if it is in `-` namespace (and will probably move in `M`), is somehow a metadata. A call to `get_metadata('favicon')` should work and return a bytes.

> Also, should we provide this dict under another name?

I'm a bit disturbed with this dictionary. There are some mandatory keys in the specs but the implementation never enforce this. If we go this way, it should be in libzim itself, not in the wrapper. And if we don't, I'm not sure the wrapper should enforce this.

I don't know what is the best here (about encoding and dict). But we should think about this :)

Reply:
OK, I am also quite uncomfortable with the `MANDATORY_METADATA_KEYS` and `Creator.mandatory_metadata_ok`. Can we agree this is a wrong interpretation of the spec (nothing's actually mandatory) and remove both?

Thanks for the additional clarification on future lib's `get_metadata()`.
I think it's a bit cumbersome to have to convert to bytes then decode manually to get a key/value string but I don't like magic decoding so I think we should just mirror get_article for now (and thus revert and fix the docstring as impl. was ok).

But this highlights the need to query existing metadata and there is no way in the current implementation to get the list of M/ articles. I can get the number of articles but that's it. I can not discover metadata and that's problematic.
Is there a way in the libzim to do it that's not exposed in this impl? The only way I see for now is to decrement-loop on article count (because M/ are stored before X/) and test urls.

Reply:
> OK, I am also quite uncomfortable with the MANDATORY_METADATA_KEYS and Creator.mandatory_metadata_ok. Can we agree this is a wrong interpretation of the spec (nothing's actually mandatory) and remove both?

There are some mandatory metadata in (https://www.openzim.org/wiki/Metadata). It was somehow a grey area between openzim and kiwix specification. But whatever how we interpret it, it must be consistent between c++ and python. I think we can remove it from python wrapper. If we need it, we must add it on c++ side.

> I think it's a bit cumbersome to have to convert to bytes then decode manually to get a key/value string but I don't like magic decoding so I think we should just mirror get_article for now (and thus revert and fix the docstring as impl. was ok).

We could at least convert it to `bytes` (instead of `memoryview`) as we know that the size is reasonable.

> But this highlights the need to query existing metadata and there is no way in the current implementation to get the list of M/ articles. I can get the number of articles but that's it. I can not discover metadata and that's problematic.
Is there a way in the libzim to do it that's not exposed in this impl? The only way I see for now is to decrement-loop on article count (because M/ are stored before X/) and test urls.

You are right. I haven't think about that but we definitively need to make the metadata discoverable. I will take this into account in the new API.